design/open/staging-compaction.md

```diff
@@ -0,0 +1,172 @@
+# Proposal: Staging Compaction
+
+## Problem Description
+
+lakeFS staging area is built on top of the [KV store](../accepted/metadata_kv/index.md).
+In short, the structure for the `Branch` entity in the kv-store is as follows:
+```go
+type Branch struct {
+	CommitID     CommitID
+	StagingToken StagingToken
+	// SealedTokens - Staging tokens are appended to the front, this allows building the diff iterator easily
+	SealedTokens []StagingToken
+}
+```
+
+Uncommitted entries are read from the staging token first, then from the 
+sealed tokens by order. Writes are performed on the staging token.
+
+The KV design has proven to meet lakeFS requirements for its consistency 
+guarantees. For most use-cases of uncommitted areas in lakeFS, it does that 
+efficiently. However, there are some cases where it falls short, for example 
+this [issue](https://github.com/treeverse/lakeFS/issues/2092).  
+There might be other cases where the structure of the `Branch` entity impacts 
+the performance of reading from the staging area, for example when the 
+number of sealed tokens is large (N)[^1] and reading a missing entry requires 
+reading from all N sealed tokens. 
```

Comment:
As far as I know multiple sealed tokens case is rare.
Is there an open open performance issue where we found more than one token that cause performance issue?

Reply:
Nope.  Well, I did see this when experimenting with very many concurrent merges.  But there are other reasons why those are slow.  Still, once I get to split staging tokens this might become more common.
Definitely something for later, agree with @itaiad200 that initially it least we should trigger this due to multiple deletions -- it will definitely help there immediately.

Reply:
It's indeed rare as every sealed token represents a fail operation on the branch, so I don't think it has a big impact.